arcgis-ios-sdk-samples/Utility network/Create load report/CreateLoadReportViewController.swift

```diff
@@ -0,0 +1,268 @@
+        let deviceTableName = "Electric Distribution Device"
+        let assetGroupName = "Circuit Breaker"
+        let assetTypeName = "Three Phase"
+        let terminalName = "Load"
+        let globalID = UUID(uuidString: "1CAF7740-0BF4-4113-8DB2-654E18800028")!
+        
+        // Create a default starting location.
+        let networkSource = utilityNetwork.definition.networkSource(withName: deviceTableName)
+        if let assetType = networkSource?.assetGroup(withName: assetGroupName)?.assetType(withName: assetTypeName),
+           let startingLocation = utilityNetwork.createElement(with: assetType, globalID: globalID),
+           let terminal = assetType.terminalConfiguration?.terminals.first(where: { $0.name == terminalName }) {
+            // Set the terminal for the location. (For our case, use the "Load" terminal.)
+            startingLocation.terminal = terminal
+            return startingLocation
+        }
+        return nil
+    }
+    
+    /// Get the utility tier's trace configuration.
+    func getTraceConfiguration() -> AGSUtilityTraceConfiguration? {
+        // Constants for creating the default trace configuration.
+        let domainNetworkName = "ElectricDistribution"
+        let tierName = "Medium Voltage Radial"
+        // Get a default trace configuration from a tier to update the UI.
+        let domainNetwork = utilityNetwork.definition.domainNetwork(withDomainNetworkName: domainNetworkName)
+        let utilityTierConfiguration = domainNetwork?.tier(withName: tierName)?.traceConfiguration
+        return utilityTierConfiguration
+    }
+    
+    /// Add a default phase to the list to better showcase the sample.
+    func addDefaultPhase() {
+        guard let defaultPhase = phaseChoices.first else { return }
+        phaseSummaries.append((defaultPhase, nil))
+        tableView.insertRows(at: [IndexPath(row: phaseSummaries.endIndex - 1, section: 0)], with: .automatic)
+    }
+    
+    // MARK: Actions
+    
+    @IBAction func resetBarButtonItemTapped(_ sender: UIBarButtonItem) {
+        phaseSummaries.removeAll()
+        tableView.reloadData()
+    }
+    
+    @IBAction func addBarButtonItemTapped(_ sender: UIBarButtonItem) {
+        let selectedPhases = Set(phaseSummaries.map { $0.phase })
+        let remainingPhases = phaseChoices.filter { !selectedPhases.contains($0) }
+        guard !remainingPhases.isEmpty else { return }
+        let alertController = UIAlertController(
+            title: "Add a phase to get the load report.",
+            message: nil,
+            preferredStyle: .actionSheet
+        )
+        remainingPhases.forEach { phase in
+            alertController.addAction(UIAlertAction(title: phase.name, style: .default) { _ in
+                self.phaseSummaries.append((phase, nil))
+                self.tableView.insertRows(at: [IndexPath(row: self.phaseSummaries.endIndex - 1, section: 0)], with: .automatic)
```

Comment:
Have you thought about keeping the table view sorted by name? As far as I know, the Swift standard library doesn't provide a way to get the insertion index into a sorted array, but `Foundation` does:

```swift
let insertionIndex = (selectedPhases as NSArray)
    .index(of: phase, inSortedRange: NSRange(selectedPhases.indices), options: [.insertionIndex], usingComparator: { ($0 as! AGSCodedValue).name.compare(($1 as! AGSCodedValue).name) })
selectedPhases.insert(phase, at: insertionIndex)
self.tableView.insertRows(at: [IndexPath(row: insertionIndex, section: 0)], with: .automatic)
```

It may not look pretty, but it gets the job done. Just a thought.

Reply:
IMO the order doesn't matter too much

- I tend to keep the list of phases in the order of how we added them (sort by date added 😆 )
- I'm not sure if it makes sense to sort current phases by names... 😅  But I sorted the list of choices/UIActions because it's easier for user to pick
- The original csharp design doesn't sort the list

![image](https://user-images.githubusercontent.com/9660181/110678020-af36b480-818a-11eb-99c5-67dd0a28bc9f.png)

Reply:
I don't care very much how the list is sorted, but it seems like it would be nice to have the order be sorted somehow, so that the order is always the same. Not to completely redesign this sample for iOS, but another options is to do away with the action sheet and have two sections in the table view, one with the selected phases and one with the available ones and let users order them however they want. See Settings > [Control Center](https://support.apple.com/en-us/HT211812) for a good example of that.

Reply:
And it is very interesting to learn the ObjC way of saying binary search 🤣 Never heard of this method before. Thanks for sharing!

Ref: https://github.com/lionheart/openradar-mirror/issues/23